### PR TITLE
Link number actions instead of embedding them

### PR DESCRIPTION
**Pull request by matt-gardner**
This PR depends on #915.  I moved the number handling logic into the `KnowledgeGraph` code, and I made it so the model can use three entity types instead of two (cells, columns, and numbers).  Now number actions are treated as knowledge graph entity productions, so they go through our linking logic instead of our embedding logic.

While I was at it, I handled some of the more common missing numbers that we ran into in the actual data - we now correctly map things like "7th", "1ghz", "7 million" and "october" to the right numbers (7, 1, 7000000, and 10, respectively).

### CODE REVIEW

**Comment by matt-gardner** [comment]
Ok, this should be ready for review.  It looks a lot bigger than it is, because I updated some vocab files and changed a python filename.

File: allennlp/data/semparse/knowledge_graphs/table_question_knowledge_graph.py [hunk 1]
```diff
@@ -0,0 +1,252 @@
...
+            number: Union[int, float] = None
+            token_text = token.text
+            text = token.text.replace(',', '').lower()
+            if text in NUMBER_WORDS:
+                number = NUMBER_WORDS[text]
+
+            magnitude = 1
+            if i < len(tokens) - 1:
+                next_token = tokens[i + 1].text.lower()
+                if next_token in ORDER_OF_MAGNITUDE_WORDS:
```
[the remark this answers is not in the document]
Comment (matt-gardner) [reply allennlp/data/semparse/knowledge_graphs/table_question_knowledge_graph.py]: I'm not sure.  PNP relies on CoreNLP to detect numbers and their values, so if CoreNLP deals with it, then PNP also will.  It was just a relatively frequent error type that I saw in the data processing logs, so I decided to handle it.

File: allennlp/models/encoder_decoders/wikitables_semantic_parser.py [hunk 2]
```diff
@@ -475,49 +482,46 @@ def _get_linking_probabilities(worlds: List[WikiTablesWorld],
...
+                # We're subtracting one here because of the null entity we added above.
+                num_entities_in_instance += len(entity_indices) - 1
+
+                # We separate the scores by type, since normalization is done per type.  There's an
+                # extra "null" entity per type, also, so we have `num_entities_per_type + 1`.  We're
+                # selecting from a (num_question_tokens, num_entities) linking tensor on _dimension 1_,
+                # so we get back something of shape (num_question_tokens,) for each index we're
+                # selecting.  All of the selected indices together then make a tensor of shape
+                # (num_question_tokens, num_entities_per_type + 1).
+                indices = Variable(linking_scores.data.new(entity_indices)).long()
```
[the remark this answers is not in the document]
Comment (matt-gardner) [reply allennlp/models/encoder_decoders/wikitables_semantic_parser.py]: I'm thinking we should punt on centralizing all of that stuff until pytorch 0.4 comes out, merging tensor and variable, and then we can figure out the best way to provide these factory methods.

File: allennlp/models/encoder_decoders/wikitables_semantic_parser.py [hunk 3]
```diff
@@ -475,49 +482,46 @@ def _get_linking_probabilities(worlds: List[WikiTablesWorld],
...
+                # But we want the null entity's score to be 0, so we set that here.
+                entity_scores[:, 0] = 0
+
+                # No need for a mask here, as this is done per batch instance, with no padding.
+                type_probabilities = torch.nn.functional.softmax(entity_scores, dim=1)
+                all_probabilities.append(type_probabilities[:, 1:])
+
+            # We need to add padding here if we don't have the right number of entities.
             if num_entities_in_instance != num_entities:
                 zeros = Variable(linking_scores.data.new(num_question_tokens,
```
[the remark this answers is not in the document]
Comment (matt-gardner) [reply allennlp/models/encoder_decoders/wikitables_semantic_parser.py]: Same as above.

File: tests/models/encoder_decoders/wikitables_semantic_parser_test.py [hunk 4]
```diff
@@ -28,40 +28,45 @@ def test_get_neighbor_indices(self):
...
         # padding of -1 is used, and correct neighbor indices.
-        assert torch.equal(neighbor_indexes.data, torch.LongTensor([[[2, 3],
-                                                                     [2, 3],
-                                                                     [0, 1],
-                                                                     [0, 1]],
-                                                                    [[1, -1],
-                                                                     [0, -1],
-                                                                     [-1, -1],
-                                                                     [-1, -1]]]))
+        print(neighbor_indices.data.numpy())
```
[the remark this answers is not in the document]
Comment (matt-gardner) [reply tests/models/encoder_decoders/wikitables_semantic_parser_test.py]: Thanks, done.

File: tests/models/encoder_decoders/wikitables_semantic_parser_test.py [hunk 5]
```diff
@@ -77,28 +82,31 @@ def test_get_linking_probabilities(self):
...
-        #     that number will approach 2 as the unnormalized linking scores for each entity get
-        #     higher.
-        true_probability = [[[0.5761169, 0.2119416, 0.0058998, 0.8756006],
-                             [0.0024561, 0.9908675, 0.0008947, 0.9811352]],
-                            [[0.7310586, 0.9996647, 0.0, 0.0],
-                             [0.0, 0.0, 0.0, 0.0]]]
+        #     null entity.  We have three entity types here, so each row should sum to at most 3,
+        #     and that number will approach 3 as the unnormalized linking scores for each entity
+        #     get higher.
+        print(entity_probability.data.cpu().numpy())
```
[the remark this answers is not in the document]
Comment (matt-gardner) [reply tests/models/encoder_decoders/wikitables_semantic_parser_test.py]: Thanks, done.

File: tests/models/encoder_decoders/wikitables_semantic_parser_test.py [hunk 6]
```diff
@@ -477,11 +485,13 @@ def test_get_entity_action_logits(self):
         assert_almost_equal(action_logits[2, 0].data.cpu().numpy(), 1.0 * .3 + 1.1 * .3 + 1.2 * .4)
 
         embedding_matrix = decoder_step._entity_type_embedding.weight.data.cpu().numpy()
-        assert_almost_equal(type_embeddings[0, 0].data.cpu().numpy(), embedding_matrix[1])
+        print(embedding_matrix)
+        print(type_embeddings)
```
[the remark this answers is not in the document]
Comment (matt-gardner) [reply tests/models/encoder_decoders/wikitables_semantic_parser_test.py]: Thanks, done.

File: allennlp/models/encoder_decoders/wikitables_semantic_parser.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (pdasigi) [reply allennlp/models/encoder_decoders/wikitables_semantic_parser.py]: Sounds good.

**Pull request merged**